cur-lib/cur/stdlib/pattern-tree.rkt

```diff
@@ -0,0 +1,699 @@
+; its use.
+; However, it might assume it's running the macro-expander to have access to
+; type information.
+; If so, it should remain in a begin-for-syntax.
+(begin-for-syntax
+  ;; Pattern tree structures; see NESTED grammar above.
+  (struct pt-decl (match-var matches) #:transparent)
+  (struct pt-match (pattern decl-or-body) #:transparent)
+  (struct pt-body (value) #:transparent)
+
+  ;; C-group; represents match data for a constructor and the associated pattern
+  ;; sub-matrix
+  (struct C-group
+    (head-patterns [pattern-sub-matrix #:mutable] bodies temporaries-map
+                   is-pattern-variable? sort-order)
+    #:transparent)
+
+  ;; For an input of form
+  ;; ((n m)
+  ;;  ([z z => A]
+  ;;   [z (s x) => B])
+  ;;
+  ;; We create the following resulting tree:
+  ;; (pt-decl n (list (pt-match z (pt-decl m (list (pt-match z A)
+  ;;                                               (pt-match (s x) B))))
+  ;; Note: if no type environment is provided, all inputs are treated as either
+  ;; pattern variables or exact syntax matches.
+  (define (create-pattern-tree stx #:env [env '()])
+    (syntax-parse stx
+      [((match-vars:id ...) ((pattern-matrix ... (~datum =>) bodies) ...))
+       #:fail-unless (for/and ([pattern-row (attribute pattern-matrix)])
+                       (= (length pattern-row) (length (attribute match-vars))))
+       "expected all match cases to have same number of patterns as number of matching variables"
+       #:fail-unless (not (zero? (length (attribute match-vars))))
+       "expected at least one matching variable"
+       (create-pattern-tree-decl-helper
+        ; HACK PR103: this map lets us can keep track of which match vars were not
+        ; generated
+        (map (lambda (mv) (syntax-property mv 'is-pos-arg? #t))
+             (attribute match-vars))
+        (attribute pattern-matrix)
+        (attribute bodies)
+        env)]))
+
+  ;; Create a prefix tree by keying on the constructor type of the first
+  ;; pattern.
+  ;;
+  ;; Conceptual step-by-step example:
+  ;; 1. (x y z) ([a b c => body1]
+  ;;             [d e f => body2])
+  ;;
+  ;; 2. (y z) (x [a ([b c => body1])]
+  ;;             [d ([e f => body2])]))
+  ;;
+  ;; 3. (z) (x [a (y [b ([c => body1])])]
+  ;;           [d (y [e ([f => body2])])]))
+  ;;
+  ;; 4. () (x [a (y [b (z [c (body1)])])]
+  ;;          [d (y [e (z [f (body2)])])]))
+  ;;
+  ;; Nested constructor patterns example: (x) ([(C1 (C2 a1)) => body])
+  ;; We rewrite this as (x temp1) ([(C1 temp1) (C2 a1) => body])
+  ;; before proceding as shown above.
+  ;; Observe that the bindings for the new temporaries are reversed; this is
+  ;; addressed in the matching function.
+  (require racket/dict)
+  (define (create-pattern-tree-decl-helper match-vars pattern-matrix bodies env)
+    (let* (; create a hash to store results for the leading pattern column
+           ; TODO PR103: Is this hash purely an optimization? Looks like an
```

Comment:
The hash is to determine uniqueness, I think the idea is that we want to merge keys at this step.
For example, if two match cases start with the same match expression then they should be merged into a single node.